packages/unmock-core/src/generator-experimental.ts

```diff
@@ -0,0 +1,328 @@
+
+export const useIfHeaderLastMile = (p: Parameter, r: Option<Schema>): Option<[string, Schema]> =>
+  isNone(r) ? none : some([p.name, r.value || { type: "string" }]);
+
+export const useIfHeader = (o: OpenAPIObject, p: Parameter): Option<[string, Schema]> =>
+  p.in !== "header"
+    ? none
+    : useIfHeaderLastMile(
+      p,
+      p.schema
+        ? isReference(p.schema)
+          ? getSchemaFromRef(o, refName(p.schema))
+          : some(p.schema)
+        : some({ type: "string"}));
+
+export const identityGetter = <T>() => new Getter<T, T>(i => i);
+
+export const parameterSchema = (o: OpenAPIObject) => new Optional<Parameter, [string, Schema]>(
+  a => useIfHeader(o, a),
+  a => s => ({ ...s, name: a[0], schema: a[1] }),
+);
+
+const cutPath = (paths: string[], path: string): string =>
+  paths.length === 0
+    ? path
+    : path.slice(0, paths[0].length) === paths[0]
+      ? path.slice(paths[0].length)
+      : cutPath(paths.slice(1), path);
+
+const removeTrailingSlash = (s: string) => s.length === 0 ? s : s.slice(-1) === "/" ? s.slice(0, -1) : s;
+
+export const truncatePath = (path: string, o: OpenAPIObject, i: ISerializedRequest) =>
+  cutPath(matchUrls(i.protocol, i.host, o).map(u => removeTrailingSlash(new url.URL(u).pathname)), path);
+
+// const trace = <T>(t: T, m?: string): T => { console.log("TRACE", t, m); return t };
+export const matcher = (req: ISerializedRequest, r: Record<string, OpenAPIObject>): Record<string, OpenAPIObject> =>
+  objectToArray<OpenAPIObject>()
+  .composeTraversal(fromTraversable(array)())
+  .composeLens(valueLens())
+  .modify(oai => Optional.fromNullableProp<OpenAPIObject>()("paths")
+    .composeIso(objectToArray<PathItem>())
+    .composeTraversal(fromTraversable(array)())
+    .composeLens(valueLens())
+    .modify(pathItem => prunePathItem(req.method, pathItem)) (
+        {
+          ...oai,
+          ...(oai.paths ? {
+              paths: Object.entries(oai.paths)
+                .reduce((i, [n, o]) =>
+                  ({ ...i, ...(matches(truncatePath(req.pathname, oai, req), n) ? {[n]: o} : {})}), {}),
+              } : {}),
+        }))(
+          Object.entries(r)
+            .reduce((i, [n, o]) =>
+              ({ ...i, ...(matchUrls(req.protocol, req.host, o).length > 0 ? {[n]: o} : {})}), {}));
+
+export const hoistTransformer = (f: (req: ISerializedRequest, r: OpenAPIObject) => OpenAPIObject) =>
+    (req: ISerializedRequest, r: Record<string, OpenAPIObject>): Record<string, OpenAPIObject> =>
+  objectToArray<OpenAPIObject>()
+  .composeTraversal(
+    fromTraversable(array)<[string, OpenAPIObject]>()
+      .filter(([__, o]) =>
+        matchUrls(req.protocol, req.host, o).length > 0))
+  .composeLens(valueLens()).modify(oai => f({
+    ...req,
+    path: truncatePath(req.path, oai, req),
+    pathname: truncatePath(req.pathname, oai, req),
+  }, oai))(r);
+
+export function responseCreatorFactory2({
```

Comment:
This is a very long function, would it be possible to split this into smaller logical units?

Reply:
Ouch, yeah, it is quite long. Hm, difficult to know what the logical units would be, as everything kind of builds on everything else...

One thing that is possible is making all of those lenses functions that create a lens. That'd reduce the size by around 60-70 lines, but perhaps at the expense of readability. Not sure what the best way forward is.